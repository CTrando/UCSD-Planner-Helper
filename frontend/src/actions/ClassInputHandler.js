import {
    setConflicts,
    setCourseNum, setDepartment, setEditOccurred,
    setInstructor, setInstructors,
    setPriority, setTypes
} from "./ClassInputMutator";
import {addClass, editClass, enterInputMode, populateSectionData, removeClass} from "./ClassInputActions";
import {setUID} from "./ScheduleGenerationActions";
import {setInstructorPref, setPriorityPref} from "./SchedulePreferenceHandler";

/**
 * Is responsible for handling all ClassInput actions, which includes running business logic when changing fields to adding
 * and removing classes
 */
export class ClassInputHandler {
    constructor(dispatch, getState) {
        this.dispatch = dispatch;
        this.state = getState().ClassInput;
        this.getState = getState;
    }

    /**
     * Edits the department and runs business rule validation on it
     */
    onDepartmentChange(rawDepartment) {
        let department = rawDepartment.trim();
        // sets the department in the store
        this.dispatch(setDepartment(department));

        // don't alter anything if the department isn't even valid
        if(!this.state.departments.includes(department))
            return;

        // no reason to change if the two are the same
        if(this.state.department === department)
            return;

        // set all other fields to blank
        this.dispatch(setCourseNum(null));
        this.dispatch(setInstructor(null));
        // will autoconvert to empty list
        this.dispatch(setConflicts(null));
        this.dispatch(setPriority(null));

        // populate the course nums and the data
        this.dispatch(populateSectionData(department));

        if (this.state.editMode)
            this.dispatch(setEditOccurred(true));
    }

    onCourseNumChange(rawCourseNum) {
        let courseNum = rawCourseNum.trim();
        this.dispatch(setCourseNum(courseNum));

        if (!this.state.courseNums.includes(courseNum))
            return;

        // must clear out the fields
        this.dispatch(setInstructor(null));
        this.dispatch(setPriority(null));
        this.dispatch(setConflicts(null));

        // record the edit
        if (this.state.editMode)
            this.dispatch(setEditOccurred(true));

        const instructors = this.state.instructorsPerClass[courseNum];
        if (!instructors)
            console.warn(`Instructors are undefined for course num ${courseNum}`);

        const types = this.state.classTypesPerClass[courseNum];
        if (!types)
            console.warn(`Class Types are undefined for course num ${courseNum}`);

        this.dispatch(setTypes(types));
        this.dispatch(setInstructors(instructors));
    }

    onInstructorChange(rawInstructor) {
        let instructor = rawInstructor.trim();

        console.log(this.state.editMode);
        // record the edit
        if (this.state.editMode)
            this.dispatch(setEditOccurred(true));

        this.dispatch(setInstructor(instructor));
    }

    onConflictChange(conflicts) {
        // record the edit
        if (this.state.editMode)
            this.dispatch(setEditOccurred(true));

        this.dispatch(setConflicts(conflicts));
    }

    onPriorityChange(priority) {
        // record the edit
        if (this.state.editMode)
            this.dispatch(setEditOccurred(true));

        this.dispatch(setPriority(priority));
    }

    buildClassFromInput() {
        // should refactor this into ClassSkeleton
        let newClass = {};
        newClass['classTitle'] = `${this.state.department} ${this.state.courseNum}`;
        newClass['courseNum'] = this.state.courseNum;
        newClass['department'] = this.state.department;
        newClass['instructor'] = this.state.instructor;
        newClass['conflicts'] = this.state.conflicts;

        return newClass;
    }

    autosave() {
        if (this.state.editMode && this.state.editOccurred) {
            // just save everything
            let newClass = this.buildClassFromInput();

            console.log(this.state);
            console.log(newClass);
            this.dispatch(setEditOccurred(false));
            this.dispatch(editClass(this.state.id, newClass));
        }
    }

    handleEdit() {
        this.autosave();
        this.dispatch(enterInputMode());
    }

    handleRemove() {
        // show an error message saying what class was dropped if that class
        // is a valid one
        if (this.state.departments.includes(this.state.currentDepartment) &&
            this.state.courseNums.includes(this.state.currentCourseNum)) {
            this.state.messageHandler.showSuccess(`Removed class ${this.state.currentDepartment} 
                                    ${this.state.currentCourseNum}`, 1000);
        } else {
            console.log("removing here");
            this.dispatch(removeClass(this.state.id));
            this.state.messageHandler.showSuccess("Successfully removed class", 1000);
        }

        this.dispatch(enterInputMode());
    }

    isDuplicate(newClass) {
        // testing whether this is a duplicate class
        return Object.values(this.getState().ClassList.selectedClasses).reduce(function (accumulator, previousClass) {
            return accumulator || newClass.classTitle === previousClass['classTitle']
        }, false);
    }

    handleAdd() {
        // gotta have course num and department to do anything
        // checking for both null and undefined
        if (!this.state.courseNum || !this.state.department)
            return;

        let error = false;
        let newClass = this.buildClassFromInput();

        if (this.isDuplicate(newClass)) {
            this.state.messageHandler.showError(`Class ${this.state.currentDepartment} ${this.state.currentCourseNum} has already been added!`);
            return;
        }

        // error checking on department and course num
        if (!this.state.departments.includes(this.state.department))
            error = true;
        if (!this.state.courseNums.includes(this.state.courseNum))
            error = true;

        if (error) {
            this.state.messageHandler.showError(`Failed to add class ${this.state.department} ${this.state.courseNum}`, 1000);
            return;
        }

        const classTitle = `${this.state.department} ${this.state.courseNum}`;
        const instructor = this.state.instructor;
        const priority = this.state.priority;

        // using the addClass method from the reducer
        this.dispatch(addClass(newClass));

        // adding preferences
        this.dispatch(setInstructorPref(classTitle, instructor));
        this.dispatch(setPriorityPref(classTitle, priority));

        this.dispatch(setInstructor(null));
        this.dispatch(setCourseNum(null));
        this.dispatch(setPriority(null));
        this.dispatch(setConflicts(null));

        this.dispatch(setUID(null));
    }
}

export function getInputHandler() {
    return function (dispatch, getState) {
        return new ClassInputHandler(dispatch, getState);
    }
}