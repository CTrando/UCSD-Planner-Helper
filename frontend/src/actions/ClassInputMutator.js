export const SET_INSTRUCTORS = "SET_INSTRUCTORS";
export const SET_DEPARTMENTS = "SET_DEPARTMENTS";
export const SET_COURSE_NUMS = "SET_COURSE_NUMS";
export const SET_TYPES = "SET_TYPES";

export const SET_DEPARTMENT = "SET_DEPARTMENT";
export const SET_COURSE_NUM = "SET_COURSE_NUM";
export const SET_INSTRUCTOR = "SET_INSTRUCTOR";
export const SET_EDIT_OCCURRED = "SET_EDIT_OCCURRED";
export const SET_CONFLICTS = "SET_CONFLICTS";
export const SET_PRIORITY = "SET_PRIORITY";
export const SET_EDIT_MODE = "SET_EDIT_MODE";
export const SET_MESSAGE_HANDLER = "SET_MESSAGE_HANDLER";


export function setCourseNums(courseNums) {
    return {
        type: SET_COURSE_NUMS,
        courseNums: courseNums
    }
}

export function setDepartments(departments) {
    return {
        type: SET_DEPARTMENTS,
        departments: departments
    }
}

export function setTypes(types) {
    return {
        type: SET_TYPES,
        types: types
    }
}

export function setInstructors(instructors) {
    return {
        type: SET_INSTRUCTORS,
        instructors: instructors,
    }
}

export function setDepartment(department) {
    return {
        type: SET_DEPARTMENT,
        department: department
    }
}

export function setCourseNum(courseNum) {
    return {
        type: SET_COURSE_NUM,
        courseNum: courseNum
    }
}

export function setInstructor(instructor) {
    return {
        type: SET_INSTRUCTOR,
        instructor: instructor
    }
}

export function setPriority(priority) {
    return {
        type: SET_PRIORITY,
        priority: priority
    }
}

export function setConflicts(conflicts) {
    return {
        type: SET_CONFLICTS,
        conflicts: conflicts
    }
}

export function setEditMode(uid, mode) {
    return {
        type: SET_EDIT_MODE,
        editMode: mode,
        editUID: uid
    }
}

export function setMessageHandler(messageHandler) {
    return {
        type: SET_MESSAGE_HANDLER,
        messageHandler: messageHandler
    }
}

export function setEditOccurred(occurred) {
    return {
        type: SET_EDIT_OCCURRED,
        occurred: occurred
    }
}




