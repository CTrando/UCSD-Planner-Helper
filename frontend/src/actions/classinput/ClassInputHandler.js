import {
    setClassTypesToIgnore,
    setCourseNum,
    setDepartment,
    setEditOccurred,
    setID,
    setInstructor,
    setInstructors,
    setPriority,
    setTypes
} from "./ClassInputMutator";
import {addClass, editClass, enterInputMode, populateSectionData, removeClass} from "./ClassInputActions";
import {SchedulePreferenceInputHandler} from "../schedulepreference/SchedulePreferenceInputHandler";
import {ignoreClassTypes} from "../ignoreclasstypes/IgnoreClassTypesActions";

/**
 * Is responsible for handling all ClassInput actions, which includes running business logic when changing fields to adding
 * and removing classes
 */
export class ClassInputHandler {
    constructor(dispatch, getState) {
        this.dispatch = dispatch;
        this.getState = getState;
    }

    /**
     * Edits the department and runs business rule validation on it
     */
    onDepartmentChange(rawDepartment, triggers = true) {
        if (!rawDepartment) {
            this.dispatch(setDepartment(null));
            return;
        }

        const state = this.getState().ClassInput;

        let department = rawDepartment.trim();

        // sets the department in the store
        this.dispatch(setDepartment(department));

        if(!triggers)
            return;

        // don't alter anything if the department isn't even valid
        if (!state.departments.includes(department))
            return;

        // set all other fields to blank
        this.dispatch(setCourseNum(null));
        this.dispatch(setInstructor(null));
        // will autoconvert to empty list
        this.dispatch(setClassTypesToIgnore(null));
        this.dispatch(setPriority(null));

        // populate the course nums and the data
        this.dispatch(populateSectionData(department));

        // already know the edit was valid if we made it this far
        if (state.editMode)
            this.dispatch(setEditOccurred(true));
    }

    onCourseNumChange(rawCourseNum, triggers = true) {
        if (!rawCourseNum) {
            this.dispatch(setCourseNum(null));
            return;
        }

        const state = this.getState().ClassInput;
        let courseNum = rawCourseNum.trim();
        this.dispatch(setCourseNum(courseNum));

        if(!triggers)
            return;

        if (!state.courseNums.includes(courseNum))
            return;

        // must clear out the fields
        this.dispatch(setInstructor(null));
        this.dispatch(setPriority(null));
        this.dispatch(setClassTypesToIgnore(null));

        // only record valid edits
        if (state.editMode)
            this.dispatch(setEditOccurred(true));

        const instructors = state.instructorsPerClass[courseNum];
        if (!instructors)
            console.warn(`Instructors are undefined for course num ${courseNum}`);

        const types = state.classTypesPerClass[courseNum];
        if (!types)
            console.warn(`Class Types are undefined for course num ${courseNum}`);

        this.dispatch(setTypes(types));
        this.dispatch(setInstructors(instructors));
    }

    onInstructorChange(rawInstructor) {
        if (!rawInstructor) {
            this.dispatch(setInstructor(null));
            return;
        }

        const state = this.getState().ClassInput;
        let instructor = rawInstructor.trim();
        this.dispatch(setInstructor(instructor));

        if (!state.instructors.includes(instructor))
            return;

        // only record valid edits
        if (state.editMode)
            this.dispatch(setEditOccurred(true));

        this.dispatch(setPriority(null));
        this.dispatch(setClassTypesToIgnore(null));
    }

    onClassTypesToIgnoreChange(classTypesToIgnore) {
        if (!classTypesToIgnore) {
            this.dispatch(setClassTypesToIgnore(null));
            return;
        }

        const state = this.getState().ClassInput;
        // record the edit
        if (state.editMode)
            this.dispatch(setEditOccurred(true));
        this.dispatch(setClassTypesToIgnore(classTypesToIgnore));
    }

    onPriorityChange(priority) {
        if (!priority) {
            this.dispatch(setPriority(null));
            return;
        }

        const state = this.getState().ClassInput;
        // record the edit
        if (state.editMode)
            this.dispatch(setEditOccurred(true));

        this.dispatch(setPriority(priority));
    }

    buildClassFromInput() {
        const state = this.getState().ClassInput;
        // should refactor this into ClassSkeleton
        let newClass = {};
        newClass['classTitle'] = `${state.department} ${state.courseNum}`;
        newClass['courseNum'] = state.courseNum;
        newClass['department'] = state.department;
        newClass['instructor'] = state.instructor;
        newClass['priority'] = state.priority;
        newClass['classTypesToIgnore'] = state.classTypesToIgnore;

        return newClass;
    }

    isValidAdd(newClass) {
        const state = this.getState().ClassInput;
        if (this.isDuplicate(newClass))
            return {
                valid: false,
                reason: `Class ${state.department} ${state.courseNum} has already been added!`
            };
        return this.isValid(newClass);
    }

    isValidEdit(newClass) {
        const state = this.getState().ClassInput;
        if (this.isDuplicate(newClass, state.id)) {
            return {
                valid: false,
                reason: `Class ${state.department} ${state.courseNum} has already been added!`
            };
        }

        return this.isValid(newClass);
    }

    /**
     * Returns whether a class is valid or not as long as the reason it is not if so
     * @param newClass the class to check
     * @returns {valid, reason}
     */
    isValid(newClass) {
        const state = this.getState().ClassInput;
        if (!state.departments.includes(newClass.department))
            return {
                valid: false,
                reason: "Department is not valid"
            };
        if (!state.courseNums.includes(newClass.courseNum))
            return {
                valid: false,
                reason: "Course Num is not valid"
            };
        if (newClass.instructor && !state.instructors.includes(newClass.instructor))
            return {
                valid: false,
                reason: "Instructor is not valid"
            };
        if (this.isDuplicate(newClass)) {

        }

        return {
            valid: true
        };
    }

    /**
     * Called in editMode such that whenever the user makes an edit, any changes are saved to the store
     */
    autosave(force = false) {
        const state = this.getState().ClassInput;
        if (force || (state.editMode && state.editOccurred)) {
            // just save everything
            let newClass = this.buildClassFromInput();

            let {valid, reason} = this.isValidEdit(newClass);
            if (!valid) {
                return `Failed to edit class. Reason: ${reason}`;
            }

            this.dispatch(setEditOccurred(false));
            this.dispatch(editClass(state.id, newClass));
            this.savePreferences();
        }
    }

    handleEdit() {
        const state = this.getState().ClassInput;
        let errMsg = this.autosave(true);
        if (errMsg)
            state.messageHandler.showError(errMsg, 1000);

        this.dispatch(enterInputMode());
    }

    /**
     * Handle what occurs when the remove button is hit during edit mode
     *
     * If the user has edited some of the fields such that the department and courseNum are no longer valid, then just
     * say remove, otherwise if the class was valid then output a message with the department and courseNum
     */
    handleRemove() {
        const state = this.getState().ClassInput;

        if (!state.editMode) {
            console.warn("Somehow, the user was able to trigger class removal without being in edit mode, breaking now.")
            return;
        }

        if (state.departments.includes(state.department) && state.courseNums.includes(state.courseNum))
            state.messageHandler.showSuccess(`Removed class ${state.department} ${state.courseNum}`, 1000);
        else state.messageHandler.showSuccess("Successfully removed class", 1000);

        this.dispatch(removeClass(state.id));
        this.dispatch(enterInputMode());
    }

    /**
     * Checks if a class is already in the store
     * @param newClass the class to check for duplicates
     * @param exclude class id to exclude
     * @returns boolean the store contains the given class already
     */
    isDuplicate(newClass, exclude = "") {
        let state = this.getState().ClassList;

        if(typeof exclude !== "string")
            console.warn(`The given key ${exclude} is of type ${typeof exclude} instead of string, continuing`);

        return Object.keys(state.selectedClasses).reduce(function (accumulator, key) {
            if (exclude === key)
                return accumulator;
            return accumulator || newClass.classTitle === state.selectedClasses[key]['classTitle']
        }, false);
    }

    handleAdd() {
        const state = this.getState().ClassInput;
        // gotta have course num and department to do anything
        // checking for both null and undefined
        if (!state.courseNum || !state.department)
            return;

        let newClass = this.buildClassFromInput();

        let {valid, reason} = this.isValidAdd(newClass);
        if (!valid) {
            state.messageHandler.showError(`Failed to add class. Reason: ${reason}`, 1000);
            return;
        }

        // using the addClass method from the reducer
        this.dispatch(addClass(newClass));
        this.savePreferences();

        // nulling out the other fields
        this.dispatch(setInstructor(null));
        this.dispatch(setCourseNum(null));
        this.dispatch(setPriority(null));
        this.dispatch(setClassTypesToIgnore(null));

        this.dispatch(setID(null));
    }

    savePreferences() {
        const state = this.getState().ClassInput;

        // TODO this is terrible gotta fix this, why create an object like this?
        let inputHandler = new SchedulePreferenceInputHandler(this.dispatch, this.getState);
        inputHandler.setClassSpecificPref();

        // sending the updates to the reducer for ignoring class types
        const classTitle = `${state.department} ${state.courseNum}`;
        this.dispatch(ignoreClassTypes(classTitle, state.classTypesToIgnore));
    }
}

export function getInputHandler(dispatch = null, getState = null) {
    // overloading so can be used by redux and outside of redux as well
    if (dispatch && getState)
        return new ClassInputHandler(dispatch, getState);

    return function (dispatch, getState) {
        return new ClassInputHandler(dispatch, getState);
    }
}