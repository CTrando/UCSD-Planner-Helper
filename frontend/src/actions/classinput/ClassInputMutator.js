export const SET_INSTRUCTORS = "SET_INSTRUCTORS";
export const SET_DEPARTMENTS = "SET_DEPARTMENTS";
export const SET_COURSE_NUMS = "SET_COURSE_NUMS";
export const SET_TYPES = "SET_TYPES";

export const SET_DEPARTMENT = "SET_DEPARTMENT";
export const SET_COURSE_NUM = "SET_COURSE_NUM";
export const SET_INSTRUCTOR = "SET_INSTRUCTOR";
export const SET_EDIT_OCCURRED = "SET_EDIT_OCCURRED";
export const SET_CLASS_TYPES_TO_IGNORE = "SET_CLASS_TYPES_TO_IGNORE";
export const SET_PRIORITY = "SET_PRIORITY";
export const SET_EDIT_MODE = "SET_EDIT_MODE";
export const SET_MESSAGE_HANDLER = "SET_MESSAGE_HANDLER";
export const SET_TRANSACTION_ID = "SET_TRANSACTION_ID";

export function setCourseNums(courseNums) {
    return {
        type: SET_COURSE_NUMS,
        courseNums: courseNums
    }
}

export function setDepartments(departments) {
    return {
        type: SET_DEPARTMENTS,
        departments: departments
    }
}

export function setTypes(types) {
    return {
        type: SET_TYPES,
        types: types
    }
}

export function setInstructors(instructors) {
    return {
        type: SET_INSTRUCTORS,
        instructors: instructors,
    }
}

export function setDepartment(department) {
    return {
        type: SET_DEPARTMENT,
        department: department
    }
}

export function setCourseNum(courseNum) {
    return {
        type: SET_COURSE_NUM,
        courseNum: courseNum
    }
}

export function setInstructor(instructor) {
    return {
        type: SET_INSTRUCTOR,
        instructor: instructor
    }
}

export function setPriority(priority) {
    return {
        type: SET_PRIORITY,
        priority: priority
    }
}

/**
 * This function updates the ClassInput UI with the class types to ignore for
 * the specific class given, not globally, that responsibility is delegated to the reducer
 * @param classTypesToIgnore
 * @returns {{type: string, conflicts: *}}
 */
export function setClassTypesToIgnore(classTypesToIgnore) {
    return {
        type: SET_CLASS_TYPES_TO_IGNORE,
        classTypesToIgnore: classTypesToIgnore
    }
}

export function setEditMode(mode) {
    return {
        type: SET_EDIT_MODE,
        editMode: mode,
    }
}

/**
 * Sets the id for the current Class Input
 * @param transactionID should be a string
 * @returns {{type: string, id: *}}
 */
export function setTransactionID(transactionID) {
    return {
        type: SET_TRANSACTION_ID,
        transactionID: transactionID
    }
}

export function setMessageHandler(messageHandler) {
    return {
        type: SET_MESSAGE_HANDLER,
        messageHandler: messageHandler
    }
}

export function setEditOccurred(editOccurred) {
    return {
        type: SET_EDIT_OCCURRED,
        editOccurred: editOccurred
    }
}




