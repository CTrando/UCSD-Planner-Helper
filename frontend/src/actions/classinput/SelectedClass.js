export default class SelectedClass {
    constructor(department, courseNum, transactionID) {
        this.department = department;
        this.courseNum = courseNum;
        this.transactionID = transactionID;
        this.classTitle = `${department} ${courseNum}`;

        this.instructor = null;
    }
}
