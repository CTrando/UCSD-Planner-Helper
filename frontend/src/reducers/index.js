import {combineReducers} from 'redux';
import ClassSelection from './ClassSelectionReducer';
import ScheduleGeneration from './ScheduleGenerationReducer';
import SchedulePreferences from './SchedulePreferencesReducer';
import ClassInput from './ClassInputReducer';

const reducers = combineReducers({
    SchedulePreferences,
    ClassSelection,
    ScheduleGeneration,
    ClassInput,
});

export default reducers;