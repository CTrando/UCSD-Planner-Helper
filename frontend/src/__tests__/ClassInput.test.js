import React from 'react';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';
import {applyMiddleware, createStore} from "redux";
import reducers from "../reducers";
import {setCourseNum, setCourseNums, setDepartment, setDepartments} from "../actions/classinput/ClassInputMutator";
import {getInputHandler as getReduxInputHandler} from "../actions/classinput/ClassInputHandler";
import {AutoComplete} from "primereact/components/autocomplete/AutoComplete";
import {mount} from 'enzyme';

function getInputHandler(store) {
    let fn = getReduxInputHandler();
    return fn(store.dispatch, store.getState);
}

// describe("ClassInput component", () => {
//     let store;
//
//     beforeEach((done) => {
//         store = createStore(reducers, applyMiddleware(thunk));
//         done();
//     });
//
//     test('Renders correctly with no props', () => {
//         const classInput = renderer.create(
//             <ClassInputContainer store={store}/>,
//         );
//         let tree = classInput.toJSON();
//         expect(tree).toMatchSnapshot();
//     });
//
//     test("Initialization", () => {
//         const classInput = mount(
//             <ClassInputContainer store={store}/>
//         );
//
//         let state = store.getState().ClassInput;
//
//         chaiExpect(state.instructors).to.have.lengthOf(0);
//         chaiExpect(state.types).to.have.lengthOf(0);
//         chaiExpect(state.transactionID).to.not.equal(null);
//     });
//
//     test('Can type in the department autocomplete and have it remember the input', () => {
//         const classInput = mount(
//             <ClassInputContainer store={store}/>
//         );
//
//         store.dispatch(setDepartments(["CSE"]));
//
//         store.dispatch(setDepartment("CS"));
//         chaiExpect(store.getState().ClassInput.department).to.equal("CS");
//
//         let department = classInput.find({id: "department"}).find(AutoComplete).instance();
//         chaiExpect(department.props.value).to.equal("CS");
//     });
//
//
//     test('Entering a valid department allows the user to enter course number', () => {
//         const classInput = mount(
//             <ClassInputContainer store={store}/>
//         );
//
//         let courseAutocomplete = classInput.find({id: "course-number"}).find(AutoComplete).instance();
//         chaiExpect(courseAutocomplete.props.disabled).to.equal(true);
//
//         store.dispatch(setDepartments(["CSE"]));
//         store.dispatch(setDepartment("CSE"));
//
//         chaiExpect(courseAutocomplete.props.disabled).to.equal(false);
//     });
//
//     test('Entering an invalid department keeps the course number autocomplete disabled', () => {
//         const classInput = mount(
//             <ClassInputContainer store={store}/>
//         );
//         store.dispatch(setDepartments(["Hello"]));
//         store.dispatch(setDepartment("CSE"));
//
//         let courseAutocomplete = classInput.find({id: "course-number"}).find(AutoComplete).instance();
//         chaiExpect(courseAutocomplete.props.disabled).to.equal(true);
//     });
//
//     test('Entering a valid course number opens up the instructor', () => {
//         const classInput = mount(
//             <ClassInputContainer store={store}/>
//         );
//         store.dispatch(setDepartments(["CSE"]));
//         store.dispatch(setCourseNums(["11"]));
//
//         store.dispatch(setDepartment("CSE"));
//         store.dispatch(setCourseNum("11"));
//
//         let instructor = classInput.find({id: "instructor"}).find(AutoComplete).instance();
//         chaiExpect(instructor.props.disabled).to.equal(false);
//     });
//
//     test('Entering an invalid course number keeps instructor autocomplete disabled', () => {
//         const classInput = mount(
//             <ClassInputContainer store={store}/>
//         );
//
//         store.dispatch(setDepartments(["CSE"]));
//         store.dispatch(setCourseNums(["11"]));
//
//         store.dispatch(setDepartment("CSE"));
//         store.dispatch(setCourseNum("12"));
//
//         let instructor = classInput.find({id: "instructor"}).find(AutoComplete).instance();
//         chaiExpect(instructor.props.disabled).to.equal(true);
//     });
//
//     /**
//      * Would write the case for irregular inputs when getting blank but cannot reproduce in a test, this is an issue with
//      * autocomplete on primereact
//      */
//     test('Getting suggestions for putting in course nums works correctly on regular inputs', () => {
//         const classInput = mount(
//             <ClassInputContainer store={store}/>
//         );
//
//         let inputHandler = getInputHandler(store);
//
//         store.dispatch(setDepartments(["CSE"]));
//         store.dispatch(setCourseNums(["11", "111", "1111", "21"]));
//
//         store.dispatch(setDepartment("CSE"));
//         inputHandler.onCourseNumChange("11");
//
//         let state = store.getState().ClassInput;
//         let departmentComponent = classInput.find(ClassInput).instance();
//         const result = departmentComponent.getCourseNumSuggestions(state.courseNum, state.courseNums);
//         chaiExpect(result).to.have.lengthOf(3);
//         chaiExpect(result).to.eql(["11", "111", "1111"]);
//     });
// });

