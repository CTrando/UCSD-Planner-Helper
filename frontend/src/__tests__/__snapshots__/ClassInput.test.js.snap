// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassInput component Renders correctly with no props 1`] = `
<div
  className="ci-container"
>
  <div
    className="ci--title"
  >
     Add Classes
  </div>
  <div
    className="content"
  >
    <div
      className="form-field"
    >
      <div
        className="input-header"
      >
         Department:
      </div>
      <span
        className="ui-autocomplete ui-widget ui-autocomplete-dd"
        id="department"
        style={null}
      >
        <input
          autoComplete="off"
          className="ui-inputtext ui-state-default ui-corner-all ui-widget ui-autocomplete-input ui-autocomplete-dd-input"
          defaultValue=""
          disabled={false}
          id={null}
          maxLength={null}
          name={null}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onContextMenu={null}
          onDoubleClick={null}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={null}
          onMouseDown={null}
          placeholder={null}
          readOnly={false}
          size={null}
          style={null}
          tabIndex={null}
          type="text"
        />
        <i
          className="ui-autocomplete-loader pi pi-spinner pi-spin"
          style={
            Object {
              "visibility": "hidden",
            }
          }
        />
        <button
          className="ui-button ui-widget ui-state-default ui-corner-all ui-autocomplete-dropdown ui-button-icon-only"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span
            className="pi pi-fw pi-caret-down ui-c ui-button-icon-left"
          />
          <span
            className="ui-button-text ui-c"
          >
            ui-btn
          </span>
        </button>
        <div
          className="ui-autocomplete-panel ui-widget-content ui-corner-all ui-input-overlay ui-shadow"
          style={
            Object {
              "maxHeight": "200px",
            }
          }
        >
          <ul
            className="ui-autocomplete-items ui-autocomplete-list ui-widget-content ui-widget ui-corner-all ui-helper-reset"
          />
        </div>
      </span>
    </div>
    <div
      className="form-field"
    >
      <div
        className="input-header"
      >
         Course Number:
      </div>
      <span
        className="ui-autocomplete ui-widget ui-autocomplete-dd"
        id="course-number"
        style={null}
      >
        <input
          autoComplete="off"
          className="ui-inputtext ui-state-default ui-corner-all ui-widget ui-autocomplete-input ui-autocomplete-dd-input ui-state-disabled"
          defaultValue=""
          disabled={true}
          id={null}
          maxLength={null}
          name={null}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onContextMenu={null}
          onDoubleClick={null}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={null}
          onMouseDown={null}
          placeholder={null}
          readOnly={false}
          size={null}
          style={null}
          tabIndex={null}
          type="text"
        />
        <i
          className="ui-autocomplete-loader pi pi-spinner pi-spin"
          style={
            Object {
              "visibility": "hidden",
            }
          }
        />
        <button
          className="ui-button ui-widget ui-state-default ui-corner-all ui-autocomplete-dropdown ui-button-icon-only ui-state-disabled"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <span
            className="pi pi-fw pi-caret-down ui-c ui-button-icon-left"
          />
          <span
            className="ui-button-text ui-c"
          >
            ui-btn
          </span>
        </button>
        <div
          className="ui-autocomplete-panel ui-widget-content ui-corner-all ui-input-overlay ui-shadow"
          style={
            Object {
              "maxHeight": "200px",
            }
          }
        >
          <ul
            className="ui-autocomplete-items ui-autocomplete-list ui-widget-content ui-widget ui-corner-all ui-helper-reset"
          />
        </div>
      </span>
    </div>
    <div
      className="ci--title-preference"
    >
       Preferences
    </div>
    <div
      className="form-field"
    >
      <div
        className="input-header"
      >
         Instructor Preference:
      </div>
      <span
        className="ui-autocomplete ui-widget ui-autocomplete-dd"
        id="instructor"
        style={null}
      >
        <input
          autoComplete="off"
          className="ui-inputtext ui-state-default ui-corner-all ui-widget ui-autocomplete-input ui-autocomplete-dd-input ui-state-disabled"
          defaultValue=""
          disabled={true}
          id={null}
          maxLength={null}
          name={null}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onContextMenu={null}
          onDoubleClick={null}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={null}
          onMouseDown={null}
          placeholder={null}
          readOnly={false}
          size={null}
          style={null}
          tabIndex={null}
          type="text"
        />
        <i
          className="ui-autocomplete-loader pi pi-spinner pi-spin"
          style={
            Object {
              "visibility": "hidden",
            }
          }
        />
        <button
          className="ui-button ui-widget ui-state-default ui-corner-all ui-autocomplete-dropdown ui-button-icon-only ui-state-disabled"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <span
            className="pi pi-fw pi-caret-down ui-c ui-button-icon-left"
          />
          <span
            className="ui-button-text ui-c"
          >
            ui-btn
          </span>
        </button>
        <div
          className="ui-autocomplete-panel ui-widget-content ui-corner-all ui-input-overlay ui-shadow"
          style={
            Object {
              "maxHeight": "200px",
            }
          }
        >
          <ul
            className="ui-autocomplete-items ui-autocomplete-list ui-widget-content ui-widget ui-corner-all ui-helper-reset"
          />
        </div>
      </span>
    </div>
    <div
      className="form-field"
    >
      <div
        className="input-header"
      >
         Ignore Class Types:
      </div>
      <div
        className="ui-listbox ui-inputtext ui-widget ui-widget-content ui-corner-all ui-state-disabled"
        id={null}
        style={null}
      >
        <div
          className="ui-listbox-list-wrapper"
        >
          <ul
            className="ui-listbox-list"
            style={null}
          />
        </div>
      </div>
    </div>
    <div
      className="form-field"
    >
      <div
        className="input-header"
      >
         Importance:
      </div>
      <div
        className="ui-rating ui-state-disabled"
        id={null}
        style={null}
      >
        <a
          onClick={[Function]}
        >
          <span
            className="ui-rating-icon pi pi-ban"
          />
        </a>
        <a
          onClick={[Function]}
        >
          <span
            className="ui-rating-icon pi pi-star-o"
          />
        </a>
        <a
          onClick={[Function]}
        >
          <span
            className="ui-rating-icon pi pi-star-o"
          />
        </a>
        <a
          onClick={[Function]}
        >
          <span
            className="ui-rating-icon pi pi-star-o"
          />
        </a>
      </div>
    </div>
    <div
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      <div
        className="form-button"
        onClick={[Function]}
      >
        <button
          className="ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only ui-state-disabled"
          disabled={true}
        >
          <span
            className="ui-button-text ui-c"
          >
            Add Class
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
